src/ring_theory/noetherian.lean

```diff
@@ -362,7 +362,7 @@ end
 open_locale classical
 /-- A module is Noetherian iff every nonempty set of submodules has a maximal submodule among them
 -/
-theorem set_has_maximal_iff_noetherian {R M} [ring R] [add_comm_group M] [module R M] : (∀(a : set $ submodule R M), a.nonempty → ∃ (M' ∈ a), ∀ (I ∈ a), M' ≤ I → I=M') ↔ is_noetherian R M :=
+theorem set_has_maximal_iff_noetherian {R M} [ring R] [add_comm_group M] [module R M] : (∀(a : set $ submodule R M), a.nonempty → ∃ (M' ∈ a), ∀ (I ∈ a), M' ≤ I → I = M') ↔ is_noetherian R M :=
```

Comment:
It looks like most of your proof doesn't rely on anything about `submodule`s beyond the fact that they're a partial order...
I think you could break off most of your proof into a separate lemma, about partial orders, perhaps show that if `P` is a partial order, then > is `well_founded` iff 
```
(∀(a : set P), a.nonempty → ∃ (M' ∈ a), ∀ (I ∈ a), M' ≤ I → I = M')
```

Reply:
Something very similar should be true for well founded relations. I proved the right direction quite easily 
```
exact λ wf a ha, ⟨well_founded.min wf a ha, well_founded.min_mem _ _ _, 
      λ I hI hminI, (lt_or_eq_of_le hminI).elim 
        (λ h, (well_founded.not_lt_min wf _ _ hI h).elim) 
        eq.symm⟩
```
I could find a proof for the left direction in mathlib, but this would be useful

Reply:
I would agree, but for now I added your code under `well_founded.has_min'` in `rel_classes`, and it has made the back half significantly shorter :+1:

Reply:
Here is a proof for the converse direction:
```lean
lemma well_founded_of_exists_max {α : Type*} [partial_order α]
  (exists_max : ∀ (p : set α), p.nonempty → ∃ m ∈ p, ∀ x ∈ p, m ≤ x → x = m) :
  well_founded ((>) : α → α → Prop) :=
⟨begin
  set counterexamples := {x : α | ¬ acc gt x},
  intro x,
  by_contra hx,
  obtain ⟨m, m_mem, hm⟩ := exists_max counterexamples ⟨x, hx⟩,
  refine m_mem (acc.intro _ (λ y y_gt_m, _)),
  by_contra hy,
  exact ne_of_lt y_gt_m (hm y hy (le_of_lt y_gt_m)).symm
end⟩
```
Could you add that to `order.rel_classes` and use it here?